Add SegWit support to importmulti

meshcollider opened the pull request:
Add support for segwit to importmulti, supports P2WSH, P2WPKH, P2SH-P2WPKH, P2SH-P2WSH. Adds a new `witnessscript` parameter which must be used for the witness scripts in the relevant situations.

Also includes some tests for the various import types.

~Also makes the change in #14019 redundant, but cherry-picks the test from that PR to test the behavior (@achow101).~

Fixes #12253, also addresses the second point in #12703, and fixes #14407

meshcollider committed:
Add SegWit support to importmulti with some ProcessImport cleanup

meshcollider committed:
Add segwit address tests for importmulti

meshcollider committed:
Fix typo in test_framework/blocktools

practicalswift commented on src/wallet/rpcdump.cpp:
```diff
@@ -892,180 +887,167 @@ static UniValue ProcessImport(CWallet * const pwallet, const UniValue& data, con
+            }
 
-                CTxDestination pubkey_dest = pubKey.GetID();
+            CTxDestination pubkey_dest = pubKey.GetID();
```
No longer needed?

practicalswift commented on src/wallet/rpcdump.cpp:
```diff
@@ -892,180 +887,167 @@ static UniValue ProcessImport(CWallet * const pwallet, const UniValue& data, con
-            if (pubKeys.size() && keys.size() == 0) {
-                const std::string& strPubKey = pubKeys[0].get_str();
+                // Generate the scripts
+                std::vector<unsigned char> vData(ParseHex(strWitnessScript));
```
Use another variable name to avoid shadowing existing local variable `vData` :-)

MeshCollider commented:
Thanks for reviewing @practicalswift, both comments addressed :)

achow101 commented on src/wallet/rpcdump.cpp:
```diff
@@ -858,32 +853,32 @@ static UniValue ProcessImport(CWallet * const pwallet, const UniValue& data, con
-        if (isP2SH && !IsHex(strRedeemScript)) {
-            throw JSONRPCError(RPC_INVALID_ADDRESS_OR_KEY, "Invalid redeem script");
+        // Force users to provide the witness script in its field rather than redeemscript
+        if (!strRedeemScript.empty() && script.IsPayToWitnessScriptHash()) {
```
Instead of checking that `strRedeemScript` is not empty, this should check that `strWitnessScript` is empty. Or it could check both: `!strRedeemScript.empty() && strWitnessScript.empty()`.

MeshCollider replied:
It should be possible to provide a P2WSH scriptPubKey without providing the witness script for it, e.g. watch only, so I don't think that's right

achow101 replied:
I think checking both covers that case.

MeshCollider replied:
Checking both would mean that if witnessScript and redeemScript were both present, it would allow it, which makes no sense if its not a P2WSH address

sipa replied:
@Sjors Yes, of course, just `importaddress` it. It probably works fine without this PR even.

Sjors replied:
So we currently support importing a `3...` address as watch-only, but the wallet doesn't know or care if it's SegWit or not. The new code here is for a scenario where you do know it's SegWit `script.IsPayToWitnessScriptHash()`, which means you know the `strRedeemScript`, but you don't know the `strWitnessScript`. That seems a strange situation. Why would anyone give you the legacy redeem script but not the Segwit witness script?

sipa replied:
I'm very confused by your response.

I just suggested having a test that doesn't do anything but import a bech32 address without revealing the pubkey, and see if it's treated as watch-only.

sipa replied:
@Sjors I have no idea what you're talking about. This is about the case where you're importing a bech32 P2WSH address (no P2SH), without revealing the script inside - which is a totally reasonable thing if you don't care about solvability.

Sjors replied:
@sipa I see, I got confused because the deleted line `if (isP2SH`.

Sjors replied:
@sipa I see, I may have gotten confused because Github is showing `if (isP2SH && !IsHex(strRedeemScript` as the deleted line, but this thread is about an earlier code branch.
<img width="591" alt="schermafbeelding 2018-10-22 om 14 16 57" src="https://user-images.githubusercontent.com/10217/47281424-45db9880-d605-11e8-977f-3330bbd7da82.png">

achow101 commented on src/wallet/rpcdump.cpp:
```diff
@@ -858,32 +853,32 @@ static UniValue ProcessImport(CWallet * const pwallet, const UniValue& data, con
             std::vector<unsigned char> vData(ParseHex(strRedeemScript));
             CScript redeemScript = CScript(vData.begin(), vData.end());
+            CScriptID redeem_id(redeemScript);
+            CScript redeemDestination = GetScriptForDestination(redeem_id);
```
nit: snake_case

achow101 commented on src/wallet/rpcdump.cpp:
```diff
@@ -809,29 +809,24 @@ UniValue dumpwallet(const JSONRPCRequest& request)
 
         // Optional fields.
         const std::string& strRedeemScript = data.exists("redeemscript") ? data["redeemscript"].get_str() : "";
+        const std::string& strWitnessScript = data.exists("witnessscript") ? data["witnessscript"].get_str() : "";
```
nit: snake_case

achow101 commented on src/wallet/rpcdump.cpp:
```diff
@@ -892,180 +887,167 @@ static UniValue ProcessImport(CWallet * const pwallet, const UniValue& data, con
-                const std::string& strPubKey = pubKeys[0].get_str();
+                // Generate the scripts
+                std::vector<unsigned char> vData(ParseHex(strWitnessScript));
+                CScript witnessScript = CScript(vData.begin(), vData.end());
```
nit: snake_case

achow101 commented on src/wallet/rpcdump.cpp:
```diff
@@ -892,180 +887,167 @@ static UniValue ProcessImport(CWallet * const pwallet, const UniValue& data, con
+                std::vector<unsigned char> vData(ParseHex(strWitnessScript));
+                CScript witnessScript = CScript(vData.begin(), vData.end());
+                CScriptID witness_id(witnessScript);
+                CScript witnessDestination = GetScriptForDestination(WitnessV0ScriptHash(witnessScript));
```
nit: snake_case

achow101 commented on src/wallet/rpcdump.cpp:
```diff
@@ -892,180 +887,167 @@ static UniValue ProcessImport(CWallet * const pwallet, const UniValue& data, con
+        // P2PK/P2PKH/P2WPKH
+        } else if (dest.type() == typeid(CKeyID) || dest.type() == typeid(WitnessV0KeyHash)) {
+            if (keys.size() > 1 || pubKeys.size() > 1) {
+                throw JSONRPCError(RPC_INVALID_PARAMETER, "More than private key given for one address");
```
This error wouldn't make sense if only pubkeys were given.

achow101 commented on src/wallet/rpcdump.cpp:
```diff
@@ -892,180 +887,167 @@ static UniValue ProcessImport(CWallet * const pwallet, const UniValue& data, con
                 }
+                pubkey = pubkey_temp;
+            }
+            if (pubkey.size()) {
```
Instead of checking pubkey size, check validity? (e.g. use `IsValid()` or `IsFullyValid()`)

MeshCollider replied:
IsValid() just performs the same size check anyway, but yep I'll do that for the sake of clarity, I don't think we need an IsFullyValid() check here

achow101 reviewed (commented):
Concept ACK

59f8e547130fc1cb756da7d990f65445ca196f5f could be squashed into 24092b3402693d8a2c39b640db18748b168863e9

acdba74ee011bfcafc9f28b528918815b1de4920 seems to be an unrelated change

meshcollider committed:
Add release notes for importmulti segwit change

sipa commented:
@MeshCollider I don't think it's necessary to import the pubkeys involved separately anymore since #14424, and in fact that sounds very dangerous (you could be tricked into importing a 2-of-3 multisig where you have 2 of the keys, but then receiving a payment to a P2PKH of the third key, and seeing it treated towards your watch-only balance). NACK until that is fixed (or at least restricted to not have that issue, but I believe that rebasing on 14424 will be sufficient).

MeshCollider commented:
@sipa good point, fixed

sipa commented:
I think there are a few things that need fixing:
* There doesn't seem to be code for P2SH-P2WPKH (which matters at least when passing in a pubkey in that case, which won't be imported).
* The tests only cover whether `importmulti` returns true, not whether it actually imported anything. 
* The tests are restricted to watchonly and spendable cases, but no solvable-but-not-spendable ones.

meshcollider committed:
Make getaddressinfo return solvability

MeshCollider commented:
I believe I've addressed the points, thanks for the feedback so far

sipa commented on src/wallet/rpcdump.cpp:
```diff
@@ -892,103 +889,74 @@ static UniValue ProcessImport(CWallet * const pwallet, const UniValue& data, con
             }
 
-            // Import private keys.
+        // P2PK/P2PKH/P2WPKH
```
I think it may make sense to repeat the trick from the P2SH block of 'descending' into scriptPubKey/address of the redeemscript, to the P2WSH block, and then do P2PK/P2PKH/P2WPKH as a totally new case rather than an else branch. This would make the code work correctly for P2PK/P2PKH nested inside P2WSH (maybe track that you don't permit P2WPKH inside P2WSH though).

sipa commented on test/functional/wallet_importmulti.py:
```diff
@@ -458,6 +468,74 @@ def run_test (self):
                 "timestamp": "",
             }])
 
+        # Import P2WPKH address
```
There are a number more scenarios that make sense to test, I think:
* Address based import (without key/script) of P2WSH-multisig
* For all of P2WPKH, P2WSH-multisig, P2SH-P2WPKH, P2SH-P2WSH-multisig, versions with script/pubkey but no private key, and test that the result is solvable.
* P2WPKH with private key, result must be spendable (not just watchonly)
* P2WSH-multisig and P2SH-multisig with more than 1 key, where some of the keys are specified as private keys, and some as public keys.

sipa replied:
@Sjors Correct.

sipa reviewed (commented):
Concept ACK

Sjors reviewed (commented):
Concept ACK.

I tested 65ecf2d importing a watch-only bech32 address, which worked.

I also tested importing a watch-only p2sh-p2wpkh address like so: `importmulti '[{"scriptPubKey": {"address": "2NFiDQiCdRYaRUdWMerff1qSTR4EC7zSBmB"}, "timestamp": "now", "pubkeys": ["03f33112792b07e9994933287ea71069d94176225f745822e47151d39e754d957f"]}]'`. This is considered `"issolvable": false`. Once I add the private key it does know how to solve it.

Code looks good to me, but I don't pretend to understand all the edge cases involved. The tests really help illustrate it, so the more the better, see Sipa's suggestions: 
https://github.com/bitcoin/bitcoin/pull/14454/commits/480cba3425efece7704f6350c915268079675853#r225796932)

instagibbs commented on src/wallet/rpcdump.cpp:
```diff
@@ -892,103 +886,95 @@ static UniValue ProcessImport(CWallet * const pwallet, const UniValue& data, con
-                    CKeyID vchAddress = pubkey.GetID();
-                    pwallet->MarkDirty();
-                    pwallet->SetAddressBook(vchAddress, label, "receive");
+        // P2WSH
```
// (P2SH-)P2WSH

Makes it more clear to me that the above block changed the `script`

instagibbs replied:
And same thing with: `// P2PK/P2PKH/P2WPKH`

sipa commented on src/wallet/rpcdump.cpp:
```diff
@@ -809,29 +809,24 @@ UniValue dumpwallet(const JSONRPCRequest& request)
 static UniValue ProcessImport(CWallet * const pwallet, const UniValue& data, const int64_t timestamp) EXCLUSIVE_LOCKS_REQUIRED(pwallet->cs_wallet)
```
It's specific to importing things, though, which is the scope of rpcdump.cpp?

instagibbs commented on src/wallet/rpcdump.cpp:
```diff
@@ -858,32 +853,34 @@ static UniValue ProcessImport(CWallet * const pwallet, const UniValue& data, con
-        }
-
-        // Process. //
+        CScript original_script = script;
```
`scriptpubkey_script` more self-explanatory? "original" is context dependent.

instagibbs commented on src/wallet/rpcdump.cpp:
```diff
@@ -858,32 +853,34 @@ static UniValue ProcessImport(CWallet * const pwallet, const UniValue& data, con
-
-        // Process. //
+        CScript original_script = script;
+        CTxDestination original_dest = dest;
```
`scriptpubkey_dest` more self-explanatory? "original" is context dependent.

sipa commented:
In response to https://github.com/bitcoin/bitcoin/pull/14454#discussion_r225794109:

Yes, `importaddress`.

MeshCollider commented on src/wallet/rpcdump.cpp:
```diff
@@ -809,29 +809,24 @@ UniValue dumpwallet(const JSONRPCRequest& request)
+        if (!isScript && !(scriptPubKey.getType() == UniValue::VOBJ && scriptPubKey.exists("address"))) {
             throw JSONRPCError(RPC_INVALID_PARAMETER, "Invalid scriptPubKey");
         }
+        const std::string& output = isScript ? scriptPubKey.get_str() : scriptPubKey["address"].get_str();
```
@instagibbs what do you suggest? `script_or_address`?

Sjors commented:
@MeshCollider I'm still seeing the same behavior I described above when importing a `p2sh-p2wpkh` address using just the address and public key.

sipa commented:
@Sjors You need the redeemscript to make your example solvable (otherwise there is no way for the matcher to go from a P2SH hash to knowing what script the hash is of).

With the private key the weird-evil-but-necessary magic behaviour that causes us to automatically recognize P2WPKH and P2SH-P2WPKH payments to known private keys triggers.

Sjors commented:
Your wording suggests you're not a fan of adding that weird-evil-but-optional magic behavior to the (very common) case of having one public key?

MeshCollider commented:
@Sjors I'm not really a fan of it, this code is already edge-cased enough without adding more ifs to test for, but I'm happy to add it if you really think its worthwhile.

sipa commented:
@Sjors No, but I have suggested something else before, which would perhaps also solve your issue, namely that for all single-key schemes you can just provide the pubkey, and `importmulti` would automatically try the different schemes to see which matches your address/script.

At the very least, something for another PR, though.

Sjors commented:
I'm fine with focusing on descriptors, since those are an order of magnitude more intuitive than the current `importmulti` syntax.

MeshCollider commented:
Fixed a bug where a P2WSH multisig was not treated as IsMine Spendable because the scriptPubKey wasn't being added to the wallet, and updated the test for it.

MeshCollider commented:
Comments addressed + squashed some commits together

sipa commented on src/wallet/rpcdump.cpp:
```diff
@@ -892,103 +890,83 @@ static UniValue ProcessImport(CWallet * const pwallet, const UniValue& data, con
-            if (IsValidDestination(dest)) {
-                pwallet->SetAddressBook(dest, label, "receive");
+            // Import into the wallet
+            if (!pwallet->AddWatchOnly(witness_script, timestamp)) {
```
A small issue still: for P2SH-P2WSH it's unnecessary (and undesirable, I think) to also import the inner P2WSH script as watch-only.

It's not necessary because we're already marking the P2SH-P2WSH script as watch-only, and for solvability all we need is AddCScript, not AddWatchOnly.

It's undesirable because it will cause us to treat payments to the P2WSH address as watchonly IsMine (instead of just the P2SH-P2WSH version).

MeshCollider replied:
Good point, fixed

Sjors commented on test/functional/wallet_importmulti.py:
```diff
@@ -458,6 +477,147 @@ def run_test (self):
+        assert_equal(result[0]['success'], True)
+        address_assert = self.nodes[1].getaddressinfo(address['address'])
+        assert_equal(address_assert['iswatchonly'], True)
+        assert_equal(address_assert['solvable'], False)
```
@sipa can you remind me why a Native Pay-to-Witness-Public-Key-Hash isn't `solvable`, even though legacy Pay-to-Public-Key-Hash addresses are? Is that just to reduce "evil magic" like in the P2SH-P2PKH case?
If so, maybe add a comment in the test or in the RPC doc?

sipa replied:
In this example it's not solvable because the public key isn't included. It has nothing to do with the script/witness hashing type.

In general, solvability is very easy. Start at the scriptPubKey. Whenever it includes a script hash, we must receive the actual script, and recurse into it. Whenever you encounter a pubkey hash, we must receive the actual public key. A scriptPubKey is solvable whenever you can descend all the way to scripts and pubkeys without any hashes of unknown things.

The only exception to that is that whenever you have a private key X, we also automatically import the P2WPKH and P2SH-P2WPKH versions of it, so no separate importing is needed in that case.

Sjors replied:
> In this example it's not solvable because the public key isn't included.

OK, that makes sense. Actually I misread the [legacy test](https://github.com/Sjors/bitcoin/blob/02d553b3e6fa78aa76cd825135a496590a243595/test/functional/wallet_importmulti.py#L134-L149): it does add the public key, which why it's solvable.

Sjors reviewed (commented):
I added a few more questions inline, hopefully merely to reduce my own confusion.

My above concerns are addressed in e4a66bb, modulo two tests that @sipa requested and are either missing or I overlooked them.

> The added tests only cover whether importmulti returns true, not whether it actually imported anything.

This is done now by checking properties via `getaddressinfo`.

> The added tests are restricted to watchonly and spendable cases, but no solvable-but-not-spendable ones.

Some of the `['solvable'], True)` tests now check that `['ismine'], False`, but none check for `iswatchonly'],False]`.

> Address based import (without key/script) of P2WSH-multisig

Covered in `# P2WSH multisig address without scripts or keys` and `# P2SH-P2WSH 1-of-1 multisig + redeemscript with no private key`

> For all of P2WPKH, P2WSH-multisig, P2SH-P2WPKH, P2SH-P2WSH-multisig, versions with script/pubkey but no private key, and test that the result is solvable.

_I think this one is still missing_ for P2WSH multisig, between the `# P2WSH multisig address without scripts or keys`  and `# Same P2WSH multisig address as above, but now with witnessscript + private keys` tests?

> P2WPKH with private key, result must be spendable (not just watchonly)

Where "spendable" means `("ismine": true and "iswatchonly": false)`. This is done.

> P2WSH-multisig and P2SH-multisig with more than 1 key, where some of the keys are specified as private keys, and some as public keys.

_Missing?_

I made a commit that checks `solvable` for all legacy scenarios as well: https://github.com/Sjors/bitcoin/commit/02d553b3e6fa78aa76cd825135a496590a243595

sipa commented:
utACK e4a66bb791cc832b192f4fd780d3159b98f494e3. As @Sjors points out there are some combinations for which tests can be added, but I think that can be done later.

Please squash fixups?

MeshCollider commented:
Squashed most fixups into their respective commits, just left one which is harder to squash.

sipa commented:
utACK c11875c5908a17314bb38caa911507dc6401ec49

instagibbs commented on test/functional/wallet_importmulti.py:
```diff
@@ -458,6 +477,147 @@ def run_test (self):
+        }])
+        assert_equal(result[0]['success'], True)
+        address_assert = self.nodes[1].getaddressinfo(multi_sig_script['address'])
+        assert_equal(address_assert['solvable'], False)
```
nit: if "sigsrequired" and "ismine" aren't in results can we assert their absence here?

instagibbs commented on test/functional/wallet_importmulti.py:
```diff
@@ -458,6 +477,147 @@ def run_test (self):
+        }])
+        assert_equal(result[0]['success'], True)
+        address_assert = self.nodes[1].getaddressinfo(multi_sig_script['address'])
+        assert_equal(address_assert['solvable'], True)
```
nit: ismine?

instagibbs reviewed (approved):
utACK

promag commented on src/wallet/rpcwallet.cpp:
```diff
@@ -3538,6 +3538,7 @@ UniValue getaddressinfo(const JSONRPCRequest& request)
             "  \"address\" : \"address\",        (string) The bitcoin address validated\n"
             "  \"scriptPubKey\" : \"hex\",       (string) The hex-encoded scriptPubKey generated by the address\n"
             "  \"ismine\" : true|false,        (boolean) If the address is yours or not\n"
+            "  \"solvable\" : true|false,      (boolean) If the address is solvable by the wallet\n"
```
Could have release note.

promag commented on src/wallet/rpcdump.cpp:
```diff
@@ -892,103 +890,83 @@ static UniValue ProcessImport(CWallet * const pwallet, const UniValue& data, con
-                    pwallet->UpdateTimeFirstKey(timestamp);
-                }
+        // (P2SH-)P2PK/P2PKH/P2WPKH
+        if (dest.type() == typeid(CKeyID) || dest.type() == typeid(WitnessV0KeyHash)) {
```
First time using RTTI? Maybe it could be avoided?

BTW, didn't test but I don't see `variant::type()` in boost 1.47 documentation.

MeshCollider replied:
@promag it seems to be there: https://www.boost.org/doc/libs/1_47_0/doc/html/boost/variant.html
I could also do `dest.which() == 1 || dest.which() == 4` but that's very unclear, or define a new enum of the types, but that seemed over the top unless there's a good reason to avoid RTTI?

achow101 commented:
utACK 201451b1ca3c6db3b13f9491a81db5b120b864bb

laanwj commented:
looks good to me, adding the `solvable` to `getaddressinfo` makes a lot of sense, and tests also look good to me
utACK c11875c5908a17314bb38caa911507dc6401ec49
